tests/test_metadata_serialization.py

```diff
@@ -0,0 +1,206 @@
+# Copyright New York University and the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+""" Unit tests testing tuf/api/metadata.py classes
+serialization and deserialization.
+
+"""
+
+import json
+import sys
+import logging
+import unittest
+import copy
+
+from typing import Any, Dict, Callable, Type
+
+from tests import utils
+
+from tuf.api.metadata import (
+    Root,
+    Snapshot,
+    Timestamp,
+    Targets,
+    Key,
+    Role,
+    MetaFile,
+    TargetFile,
+    Delegations,
+    DelegatedRole,
+)
+
+logger = logging.getLogger(__name__)
+
+# DataSet is only here so type hints can be used:
+# It is a dict of name to test dict
+DataSet = Dict[str, str]
+
+# Test runner decorator: Runs the test as a set of N SubTests,
+# (where N is number of items in dataset), feeding the actual test
+# function one data item at a time
+def run_sub_tests_with_dataset(dataset: Type[DataSet]):
+    def real_decorator(function: Callable[["TestSerialization", DataSet], None]):
+        def wrapper(test_cls: "TestSerialization"):
+            for case, data in dataset.items():
+                with test_cls.subTest(case=case):
+                    function(test_cls, data)
+        return wrapper
+    return real_decorator
+
+
+class TestSerialization(unittest.TestCase):
+
+    KEY = '{"keytype": "rsa", "scheme": "rsassa-pss-sha256", \
+        "keyval": {"public": "foo"}}'
+    SIGNED_COMMON = '"spec_version": "1.0.0", "version": 1, \
+        "expires": "2030-01-01T00:00:00Z"'
+
+    valid_keys: DataSet = {
+        "all": '{"keytype": "rsa", "scheme": "rsassa-pss-sha256", \
+            "keyval": {"public": "foo"}}',
+    }
+
+    @run_sub_tests_with_dataset(valid_keys)
```

Comment:
This turned out to be a very clever decorator which avoids calling `subTest` each time and yet all functions in the class repeat exactly the same code, except for the different metadata class. It feels imbalanced. It was not worth figuring out how this decorator works to only see repeated code everywhere else.

I understand that Key.from_dict is the only function with a different signature which is very annoying and I couldn't come up with a good alternative yet too.